docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1583 @@
+must be pairwise valid upgrades of the original types.
+
+Adding a parameter in the middle of the parameter sequence leads to a
+validation error.
+
+Adding a non-optional parameter at the end of the parameter leads to a
+validation error.
+
+.. _examples-2:
+
+**Examples**
+
+Below, the template on the right is a valid upgrade of the template on
+the left. It adds an optional parameter ``x1`` at the end of the parameter
+sequence.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - .. code-block:: daml
+ 
+             template T
+                 with
+                   p : Party
+                 where
+                   signatory p
+
+     - .. code-block:: daml
+ 
+          template T
+              with
+                p : Party
+                x1 : Optional Int
+              where
+                signatory p
+ 
+Below, the template on the right is **not** a valid upgrade of the
+template on the left because it adds a new parameter ``x1`` before ``p`` instead
+of adding it at the end of the parameter sequence.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - .. code-block:: daml
+
+            template T
+              with
+                p : Party
+              where
+                signatory p
+
+     - .. code-block:: daml
+
+            template T
+              with
+                x1 : Optional Int
+                p : Party
+              where
+                signatory p
+
+Below, the template on the right is **not** a valid upgrade of the
+template on the left because it drops parameter ``x1``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - .. code-block:: daml
+
+            template T
+              with
+                p : Party
+                x1 : Int
+              where
+                signatory p
+
+     - .. code-block:: daml
+
+            template T
+              with
+                p : Party
+              where
+                signatory p
+
+Below, the template on the right is **not** a valid upgrade of the
+template on the left because it changes the type of ``x1`` from ``Int`` to ``Text``.
+``Text`` is not a valid upgrade of ``Int``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+
+   * - .. code-block:: daml
+
+            template T
+              with
+                p : Party
+                x1 : Int
+              where
+                signatory p
+
+     - .. code-block:: daml
+
+            template T
+              with
+                p : Party
+                x1 : Text
+              where
+                signatory p
+        
+Template Keys
+~~~~~~~~~~~~~
+
+An upgrading template cannot modify the key of the upgraded template in
+any way.
+
+Adding a key leads to a validation error.
+
+Removing a key leads to a validation error.
+
+Changing the type of a key leads to a validation error.
+
+For 2.9, key types can only use definitions from the current package or 
+from ``daml-stdlib``/``daml-prim``.
```

Comment:
```suggestion
For 2.9, key types can only use definitions from the current package or 
from the Daml standard library. 
```
You could add a  ref to ``daml-stdlib``/``daml-prim``, but you need a link. It is maybe not obvious what those are.